db/corruption_test.cc

```diff
@@ -912,6 +913,258 @@ TEST_F(CorruptionTest, VerifyWholeTableChecksum) {
+ public:
+  explicit CorruptionTestWithParams1()
+      : CorruptionTest(),
+        avoid_flush_during_recovery_(std::get<0>(GetParam())),
+        track_and_verify_wals_in_manifest_(std::get<1>(GetParam())) {}
+
+ protected:
+  const bool avoid_flush_during_recovery_;
+  const bool track_and_verify_wals_in_manifest_;
+};
+
+INSTANTIATE_TEST_CASE_P(CorruptionTest, CorruptionTestWithParams1,
+                        ::testing::Values(std::make_tuple(true, false),
+                                          std::make_tuple(false, false),
+                                          std::make_tuple(true, true),
+                                          std::make_tuple(false, true)));
+
+TEST_P(CorruptionTestWithParams1, CrashDuringRecovery) {
+  CloseDb();
+  Options options;
+  options.track_and_verify_wals_in_manifest =
+      track_and_verify_wals_in_manifest_;
+  options.wal_recovery_mode = WALRecoveryMode::kPointInTimeRecovery;
+  options.avoid_flush_during_recovery = false;
+  options.env = env_;
+  ASSERT_OK(DestroyDB(dbname_, options));
+  options.create_if_missing = true;
+
+  Reopen(&options);
+  Status s;
+  const std::string test_cf_name = "test_cf";
+  ColumnFamilyHandle* cfh = nullptr;
+  s = db_->CreateColumnFamily(options, test_cf_name, &cfh);
+  ASSERT_OK(s);
+  if (track_and_verify_wals_in_manifest_) {
+    ASSERT_OK(db_->SyncWAL());
+  }
+  delete cfh;
+  CloseDb();
+
+  std::vector<ColumnFamilyDescriptor> cf_descs;
+  cf_descs.emplace_back(kDefaultColumnFamilyName, options);
+  cf_descs.emplace_back(test_cf_name, options);
+  std::vector<ColumnFamilyHandle*> handles;
+
+  ASSERT_OK(DB::Open(options, dbname_, cf_descs, &handles, &db_));
+
+  // Write one key to test_cf.
+  ASSERT_OK(db_->Put(WriteOptions(), handles[1], "old_key", "dontcare"));
+
+  // Write to default_cf and flush this cf several times to advance wal
+  // number.
+  for (int i = 0; i < 2; ++i) {
+    ASSERT_OK(db_->Put(WriteOptions(), "key" + std::to_string(i), "value"));
+    ASSERT_OK(db_->Flush(FlushOptions()));
+  }
+
+  auto* dbimpl = static_cast_with_check<DBImpl>(db_);
+  assert(dbimpl);
+
+  ASSERT_OK(db_->Put(WriteOptions(), handles[1], "dontcare", "dontcare"));
+  if (track_and_verify_wals_in_manifest_) {
+    ASSERT_OK(db_->SyncWAL());
+  }
+
+  std::vector<std::string> files;
+  ASSERT_OK(env_->GetChildren(dbname_, &files));
+  uint64_t max_wal_num = 0;
+  for (const auto& f : files) {
+    uint64_t file_num = 0;
+    FileType type{kWalFile};
+    if (ParseFileName(f, &file_num, &type) && type == kWalFile) {
+      if (file_num > max_wal_num) {
+        max_wal_num = file_num;
+      }
+    }
+  }
+
+  for (auto* h : handles) {
+    delete h;
+  }
+  handles.clear();
+  CloseDb();
+
+  std::string wal_path = LogFileName(dbname_, max_wal_num);
+  uint64_t old_size = 0;
+  ASSERT_OK(env_->GetFileSize(wal_path, &old_size));
+  assert(old_size > 8);
+  uint64_t new_size = old_size - 8;
+  ASSERT_OK(test::TruncateFile(env_, wal_path, new_size));
```

Comment:
I think it's helpful to add some comments for this code block, or even put it in a helper function.
The comments should explain what errors we are emulating: power reset which caused us to lose the last un-synced WAL.

Reply:
Same for the other unit test of transaction db.